src/core/Akka/Actor/Scheduler/DedicatedThreadScheduler.cs

```diff
@@ -141,22 +161,61 @@ private void InternalScheduleRepeatedly(TimeSpan initialDelay, TimeSpan interval
 
                     AddWork(interval, executeAction, token);
                 }
-                catch (OperationCanceledException) { }
+                catch (OperationCanceledException)
+                {
+                }
                 catch (Exception x)
                 {
-                    _log.Error(x,"DedicatedThreadScheduler failed to execute action");
+                    Log.Error(x, "DedicatedThreadScheduler failed to execute action");
                 }
             };
             AddWork(initialDelay, executeAction, token);
 
         }
 
-        private void AddWork(TimeSpan delay, Action work,CancellationToken token)
+        private void AddWork(TimeSpan delay, Action work, CancellationToken token)
         {
+            if (_stopped.Value != null)
+                throw new SchedulerException("cannot enque after timer shutdown");
             var expected = HighResMonotonicClock + delay;
-            var scheduledWord = new ScheduledWork(expected.Ticks, work,token);
+            var scheduledWord = new ScheduledWork(expected.Ticks, work, token);
             _workQueue.Enqueue(scheduledWord);
-        }               
+        }
+
+        private AtomicReference<TaskCompletionSource<bool>> _stopped = new AtomicReference<TaskCompletionSource<bool>>();
+
+        private void FireStopSignal()
+        {
+            try
+            {
+                _stopped.Value.TrySetResult(true);
+            }
+            catch (Exception)
+            {
+                
+            }
+        }
+
+        public void Dispose()
+        {
+            if (!Stop().Wait(_shutdownTimeout))
+            {    
+                Log.Warning("Failed to shutdown DedicatedThreadScheduler within {0}", _shutdownTimeout);   
+            }
+        }
+
+        private static readonly Task Completed = Task.FromResult(true);
+
+        private Task Stop()
+        {
+            var p = new TaskCompletionSource<bool>();
+            if (_stopped.CompareAndSet(null, p))
```

Comment:
Once this signal is set, any currently scheduled work will be allowed to complete 1 iteration only.